Panic and SIGSEGV for the ADOT collector for CloudWatch(AWS)

**Describe the bug**
I am getting the following issue while deploying the ADOT collector for AWS CloudWatch based on the document - https://docs.aws.amazon.com/eks/latest/userguide/configure-cw.html


kubectl logs my-collector-cloudwatch-collector-666665859f-7lrtq -n default
2022/05/12 01:17:51 AWS OTel Collector version: v0.17.0
2022/05/12 01:17:51 found no extra config, skip it, err: open /opt/aws/aws-otel-collector/etc/extracfg.txt: no such file or directory
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x8a664a]



**Steps to reproduce**
Not Reproducible. With a different EKS cluster I can't replicate the issue.

**What did you expect to see?**
I should see -
kubectl apply -f collector-config-cloudwatch.yaml
opentelemetrycollector.opentelemetry.io/my-collector-cloudwatch created
clusterrole.rbac.authorization.k8s.io/otel-prometheus-role created
clusterrolebinding.rbac.authorization.k8s.io/otel-prometheus-role-binding created

**What did you see instead?**
I saw -
2022/05/12 01:17:51 AWS OTel Collector version: v0.17.0
2022/05/12 01:17:51 found no extra config, skip it, err: open /opt/aws/aws-otel-collector/etc/extracfg.txt: no such file or directory
panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x8a664a]
goroutine 1 [running]:
go.uber.org/zap.(*Logger).check(0x0, 0x1, {0x3610ca9, 0x1b})
        go.uber.org/zap@v1.21.0/logger.go:270 +0x6a
go.uber.org/zap.(*Logger).Warn(0x0, {0x3610ca9, 0x0}, {0xc0004337c0, 0x1, 0x1})
        go.uber.org/zap@v1.21.0/logger.go:199 +0x3e
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter.(*Config).Validate(0xc0007dd180)
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter@v0.45.1/config.go:97 +0x289
go.opentelemetry.io/collector/config.(*Config).Validate(0xc0006daf20)
        go.opentelemetry.io/collector@v0.45.0/config/config.go:74 +0x2a8
go.opentelemetry.io/collector/service.(*configProvider).Get(0xc000481aa0, {0x3c16150, 0xc00012e000}, {0xc00070a3f0, 0xc00070a6c0, 0xc00070a960, 0xc00070a240})
        go.opentelemetry.io/collector@v0.45.0/service/config_provider.go:167 +0x2de
go.opentelemetry.io/collector/service.(*Collector).setupConfigurationComponents(0xc000716000, {0x3c16150, 0xc00012e000})
        go.opentelemetry.io/collector@v0.45.0/service/collector.go:171 +0xa3
go.opentelemetry.io/collector/service.(*Collector).Run(0xc000716000, {0x3c16150, 0xc00012e000})
        go.opentelemetry.io/collector@v0.45.0/service/collector.go:221 +0x24e
main.newCommand.func1(0xc000710000, {0x35d8b36, 0x1, 0x1})
        github.com/aws-observability/aws-otel-collector/cmd/awscollector/main.go:117 +0xf5
github.com/spf13/cobra.(*Command).execute(0xc000710000, {0xc000128010, 0x1, 0x1})
        github.com/spf13/cobra@v1.3.0/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc000710000)
        github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.3.0/command.go:902
main.runInteractive({{0xc00070a3f0, 0xc00070a6c0, 0xc00070a960, 0xc00070a240}, {{0x35f3e9b, 0x12}, {0x35f29a7, 0x12}, {0x3b32b10, 0x7}}, ...})
        github.com/aws-observability/aws-otel-collector/cmd/awscollector/main.go:83 +0x5d
main.run({{0xc00070a3f0, 0xc00070a6c0, 0xc00070a960, 0xc00070a240}, {{0x35f3e9b, 0x12}, {0x35f29a7, 0x12}, {0x3b32b10, 0x7}}, ...})
        github.com/aws-observability/aws-otel-collector/cmd/awscollector/main_others.go:42 +0xf8
main.main()
        github.com/aws-observability/aws-otel-collector/cmd/awscollector/main.go:76 +0x2d8



**What version did you use?**
AWS OTel Collector version: v0.17.0

**What config did you use?**
Config: Same as given in the document

**Environment**
OS: (e.g., "Ubuntu 20.04") - Node is having AMI - ami-0e4286e3300d2ec0f (amazon-eks-node-1.20-v20220406)
Compiler(if manually compiled): (e.g., "go 14.2") : No

**Additional context**
Add any other context about the problem here.

Since it was giving error in the https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/awsemfexporter/config.go validate function due to some metrics, I changed the section for exporters from -

      awsemf:
        region: "<AWS_REGION>"
        namespace: ContainerInsights/Prometheus
        log_group_name: '/aws/containerinsights/${CLUSTER_NAME}/prometheus'
        resource_to_telemetry_conversion:
          enabled: true
        dimension_rollup_option: NoDimensionRollup
        parse_json_encoded_attr_values: [Sources, kubernetes]
        metric_declarations:
          - dimensions: [[EKS_Cluster, EKS_Namespace, EKS_PodName]]
To -

      awsemf:
        region: "<AWS_REGION>"
        namespace: ContainerInsights/Prometheus
        log_group_name: '/aws/containerinsights/${CLUSTER_NAME}/prometheus'
        resource_to_telemetry_conversion:
          enabled: true

It started working. Basically removed the last 4 lines. But my query is why it is not working with the metric_declarartions?

--- a/.chloggen/fix_emfLogger.yaml
+++ b/.chloggen/fix_emfLogger.yaml
@@ -1,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: awsemfexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Ensure that the config.logger is not nil to avoid segfaults
+
+# One or more tracking issues related to the change
+issues: [9978]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:

--- a/exporter/awsemfexporter/config_test.go
+++ b/exporter/awsemfexporter/config_test.go
@@ -59,6 +59,7 @@ func TestLoadConfig(t *testing.T) {
 				LogStreamName:         "",
 				DimensionRollupOption: "ZeroAndSingleDimensionRollup",
 				OutputDestination:     "cloudwatch",
+				logger:                zap.NewNop(),
 			},
 		},
 		{
@@ -79,6 +80,7 @@ func TestLoadConfig(t *testing.T) {
 				DimensionRollupOption:       "ZeroAndSingleDimensionRollup",
 				OutputDestination:           "cloudwatch",
 				ResourceToTelemetrySettings: resourcetotelemetry.Settings{Enabled: true},
+				logger:                      zap.NewNop(),
 			},
 		},
 		{
@@ -103,6 +105,7 @@ func TestLoadConfig(t *testing.T) {
 					Unit:       "Count",
 					Overwrite:  true,
 				}},
+				logger: zap.NewNop(),
 			},
 		},
 	}

--- a/exporter/awsemfexporter/emf_exporter_test.go
+++ b/exporter/awsemfexporter/emf_exporter_test.go
@@ -545,13 +545,13 @@ func TestPushMetricsDataWithErr(t *testing.T) {
 func TestNewExporterWithoutConfig(t *testing.T) {
 	factory := NewFactory()
 	expCfg := factory.CreateDefaultConfig().(*Config)
+	settings := exportertest.NewNopCreateSettings()
 	t.Setenv("AWS_STS_REGIONAL_ENDPOINTS", "fake")
 
-	assert.Nil(t, expCfg.logger)
-	exp, err := newEmfPusher(expCfg, exportertest.NewNopCreateSettings())
+	exp, err := newEmfPusher(expCfg, settings)
 	assert.NotNil(t, err)
 	assert.Nil(t, exp)
-	assert.NotNil(t, expCfg.logger)
+	assert.Equal(t, settings.Logger, expCfg.logger)
 }
 
 func TestNewExporterWithMetricDeclarations(t *testing.T) {
@@ -633,11 +633,11 @@ func TestWrapErrorIfBadRequest(t *testing.T) {
 func TestNewEmfExporterWithoutConfig(t *testing.T) {
 	factory := NewFactory()
 	expCfg := factory.CreateDefaultConfig().(*Config)
+	settings := exportertest.NewNopCreateSettings()
 	t.Setenv("AWS_STS_REGIONAL_ENDPOINTS", "fake")
 
-	assert.Nil(t, expCfg.logger)
-	exp, err := newEmfExporter(expCfg, exportertest.NewNopCreateSettings())
+	exp, err := newEmfExporter(expCfg, settings)
 	assert.NotNil(t, err)
 	assert.Nil(t, exp)
-	assert.NotNil(t, expCfg.logger)
+	assert.Equal(t, settings.Logger, expCfg.logger)
 }

--- a/exporter/awsemfexporter/factory.go
+++ b/exporter/awsemfexporter/factory.go
@@ -19,6 +19,7 @@ import (
 
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/exporter"
+	"go.uber.org/zap"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/awsutil"
 )
@@ -47,7 +48,7 @@ func createDefaultConfig() component.Config {
 		Namespace:             "",
 		DimensionRollupOption: "ZeroAndSingleDimensionRollup",
 		OutputDestination:     "cloudwatch",
-		logger:                nil,
+		logger:                zap.NewNop(),
 	}
 }
 

--- a/exporter/awsemfexporter/factory_test.go
+++ b/exporter/awsemfexporter/factory_test.go
@@ -32,6 +32,7 @@ func TestCreateDefaultConfig(t *testing.T) {
 	cfg := factory.CreateDefaultConfig()
 	assert.NotNil(t, cfg, "failed to create default config")
 	assert.NoError(t, componenttest.CheckConfigStruct(cfg))
+	assert.NotNil(t, cfg.(*Config).logger)
 }
 
 func TestCreateTracesExporter(t *testing.T) {

--- a/exporter/awsemfexporter/grouped_metric.go
+++ b/exporter/awsemfexporter/grouped_metric.go
@@ -63,10 +63,10 @@ func addToGroupedMetric(pmd pmetric.Metric, groupedMetrics map[interface{}]*grou
 		// if patterns are provided for a valid key and that key doesn't exist in the resource attributes, it is replaced with `undefined`.
 		if !patternReplaceSucceeded {
 			if strings.Contains(metadata.logGroup, "undefined") {
-				metadata.logGroup, _ = replacePatterns(config.LogGroupName, labels, config.logger)
+				metadata.logGroup, _ = replacePatterns(config.LogGroupName, labels, logger)
 			}
 			if strings.Contains(metadata.logStream, "undefined") {
-				metadata.logStream, _ = replacePatterns(config.LogStreamName, labels, config.logger)
+				metadata.logStream, _ = replacePatterns(config.LogStreamName, labels, logger)
 			}
 		}
 

